VACMS-4850 Vet Center dashboard MVP

ethanteague opened the pull request:
## Description
See #4850

## Testing done
Visual

## Screenshots
![image](https://user-images.githubusercontent.com/2404547/114211793-12a13700-992f-11eb-8687-dbe648c9bcff.png)
![image](https://user-images.githubusercontent.com/2404547/114211807-1634be00-992f-11eb-9657-cea3fae6150a.png)
![image](https://user-images.githubusercontent.com/2404547/114211817-19c84500-992f-11eb-93d2-d6d6b403f13b.png)

## QA steps
As an administrator, go to `/taxonomy/term/190/edit` and choose the product: `Vet Centers`.
- Clear cache, and go to `/section/veterans-health-administration/vet-centers/escanaba-vet-center`
- [ ] Visually verify that you see the new dashboard format (6 cards up top, listing down below)
![image](https://user-images.githubusercontent.com/5752113/116339697-e3eae380-a7ab-11eb-88c1-0cd35f424a45.png)

- [ ]  verify view and edit links on the 6 cards work as expected.

- Go to /section/veterans-health-administration/daytona-beach-vet-center
- [ ] verify that only 5 cards appear as there is no locations page created yet for daytona. (fails gracefully)
![image](https://user-images.githubusercontent.com/5752113/116339987-65427600-a7ac-11eb-9389-7690bf072b89.png)


-  Go to a vetcenter section that has no matching node like /section/veterans-health-administration/vet-centers-mvp
![image](https://user-images.githubusercontent.com/5752113/116340774-bb63e900-a7ad-11eb-94c8-c9777b88d6be.png)


- [ ] validate that no cards appear and there are no errors  on the page (fails gracefully)
- [ ] Visit a health system section `/section/veterans-health-administration/vamc-facilities/va-salisbury-health-care` and verify that there is no dashboard and it functions as expected 
 ![image](https://user-images.githubusercontent.com/5752113/116340582-67590480-a7ad-11eb-8c27-f20015b4afbd.png)
- [ ] 
![image](https://user-images.githubusercontent.com/5752113/116340582-67590480-a7ad-11eb-8c27-f20015b4afbd.png)Visit another administration term page, and visually verify that the default view listing is displayed, e.g.: `/section/veterans-affairs`

CACHING Testing with xdebug:
- [x] Validate that the service is only called once during the initial sectionpage load
- [x] Validate that the block build() function is only called on the first page load.
- [x] Validate subsequent lods of the page do not run the service or the build(). (cached indefinitely)
- [x] Validate cache is only invalidated by cache flush OR editing he section and saving the section.


## Release instructions
- [ ] As an administrator, go to `/taxonomy/term/190/edit` and choose the product: `Vet Centers`.

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Core Application Team`
- [x] `Product Support Team`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change? 
- [ ] Yes, and it's written in issue ____ and queued for publication. 
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written 
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [ ] No announcement is needed for this code change. 
  - [ ] Merge & carry on unburdened by announcements

swirtSJW commented on docroot/modules/custom/va_gov_dashboards/src/Plugin/Derivative/VcDashboardsBlock.php:
```diff
@@ -0,0 +1,133 @@
+          ],
+        ];
+
+        $vc_locations_node_fetch = $this->entityTypeManager->getStorage('node')->loadByProperties([
+          'type' => 'vet_center_locations_list',
+          'field_administration' => $tid,
+        ]);
+        if (!empty($vc_locations_node_fetch)) {
+          $vc_locations_node = reset($vc_locations_node_fetch);
+          $vc_locations_node_id = $vc_locations_node->id();
+          $dash_panels['locations'] = [
```
It would be great to consider how we could increase the re-useabiltiy  of this.
It could be as simple as moving the things like the specific term id and the content type to reference in vars at the top in one spot and then use the var and making  the vars uses less "vet center" specific.  Or it could be a little more broad by creating a trait that provides some helpers so we don't have to have so much duplicate code in each of these dashboards.

ethanteague replied:
@swirtSJW | @kevwalsh 
Agreed - I would like to get a feel for what future dashboards are gonna look like. My thinking is, I would like to make a trait that is generic enough to allow for different panel implementations (fields, etc.). Do either of you all know what kind of stuff we are going to have in other dashboards? If this is still kinda fluid, I'm wondering if I can punt on this until we get other dash implementations a bit more nailed down?

swirtSJW replied:
Fair enough, lets make dashboard number 2 require  more abstraction of this.

swirtSJW commented on docroot/modules/custom/va_gov_dashboards/va_gov_dashboards.module:
```diff
@@ -0,0 +1,83 @@
+        break;
+
+      case '290':
+        echo 'va_gov_homepage';
```
?

ethanteague replied:
D'oh! Thanks for catching that (typo), @swirtSJW !

swirtSJW commented on docroot/modules/custom/va_gov_dashboards/va_gov_dashboards.module:
```diff
@@ -0,0 +1,83 @@
+    // If we don't have one, traverse parents.
+    if (!empty($product_id)) {
+      $parents = \Drupal::service('entity_type.manager')->getStorage("taxonomy_term")->loadAllParents($entity->id());
+      foreach ($parents as $parent) {
```
This assumes parents are returned in order of hierarchy.  Is that how they are returned?

ethanteague replied:
Yep. I just tested by throwing a breakpoint in that foreach (after removing the break statement in the loop for testing), and checking the id returned at each iteration. Went like this: `392 (Escanaba) -> 699 (Vet Centers MVP) -> 190 (Vet Centers) -> 3 (Veterans Health Administration)`

swirtSJW commented on docroot/modules/custom/va_gov_dashboards/va_gov_dashboards.module:
```diff
@@ -0,0 +1,83 @@
+    $product_id = $entity->field_product->getString() ? $entity->field_product->getString() : '';
+    // If we don't have one, traverse parents.
+    if (!empty($product_id)) {
+      $parents = \Drupal::service('entity_type.manager')->getStorage("taxonomy_term")->loadAllParents($entity->id());
```
I have some concerns about the performance of this on deep terms and how often people will be hitting the this dashboard.   This could be pretty costly.   We may want to kick around some ideas on how to cache this or create a lookup table.

ethanteague replied:
Yeah, I was kinda feeling that as well, @swirtSJW . I like the idea of a some kind of a flat file / lookup table, but not sure how extensible this would be.

ethanteague replied:
Wondering if the lookup will happen every time the page is hit, @swirtSJW ? Won't the block cache that I have setup (using the term tag) serve this as static content?

swirtSJW replied:
It should, but we need to validate that hypothisis reloading the page with the debugger running and a breakpoint set in what should not be running repeatedly

swirtSJW reviewed (commented):
Just some thoughts.

kevwalsh commented on docroot/modules/custom/va_gov_dashboards/src/Plugin/Derivative/VcDashboardsBlock.php:
```diff
@@ -0,0 +1,133 @@
+    $dash_panels = [];
+    if ($this->routeMatch->getRouteName() === 'entity.taxonomy_term.canonical' || $this->routeMatch->getRouteName() === 'layout_builder.defaults.taxonomy_term.view') {
+      $vc_node = '';
+      $tid = $this->routeMatch->getRawParameter('taxonomy_term') ? $this->routeMatch->getRawParameter('taxonomy_term') : '392';
```
392 is Escanaba Vet Center, one of 300 Vet Centers. Is that here as an example?  I didn't think TIDs would be in the code, since the trigger is in the content layer.

ethanteague replied:
That's right, @kevwalsh.

swirtSJW commented on docroot/modules/custom/va_gov_dashboards/src/Plugin/Derivative/VcDashboardsBlock.php:
```diff
@@ -0,0 +1,164 @@
+    $node = NULL;
+    $vc_node_fetch = $this->entityTypeManager->getStorage('node')->loadByProperties([
+      'type' => 'vet_center',
+      'field_administration' => $tid,
```
On line 129 you create the possiblilty that $tid might be empty,   Here would be where you want to be defensive about what should or should not happen if it is empty.

swirtSJW reviewed (changes requested):
I am interrupted on my review so it is not complete, but wanted to get this in.

swirtSJW commented on docroot/modules/custom/va_gov_dashboards/src/Plugin/Derivative/VcDashboardsBlock.php:
```diff
@@ -0,0 +1,167 @@
+   */
+  private function getCorrespondingNode(string $content_type, string $tid) {
+    if (empty($tid)) {
+      return;
```
This has to come after where you set $node = null and then it has to return $node.

swirtSJW commented on docroot/modules/custom/va_gov_dashboards/src/Plugin/Derivative/VcDashboardsBlock.php:
```diff
@@ -0,0 +1,167 @@
+  protected $entityTypeManager;
+
+  /**
+   * {@inheritdoc}
```
Because of what you did in the create, you are overriding the parameters from the base class, so you can't just inherit the doc.  So the docblock has to be explicit.

swirtSJW commented on docroot/modules/custom/va_gov_dashboards/src/Plugin/Block/VcDashboardsBlock.php:
```diff
@@ -0,0 +1,87 @@
+  protected $entityTypeManager;
+
+  /**
+   * {@inheritDoc}
```
You are altering what gets passed in via the create, so you have to explicitly declare the docblock for the constructor. It can't be inherited.

swirtSJW commented on docroot/modules/custom/va_gov_dashboards/src/Plugin/Derivative/VcDashboardsBlock.php:
```diff
@@ -0,0 +1,167 @@
+      return $node;
+    }
+    $vc_node_fetch = $this->entityTypeManager->getStorage('node')->loadByProperties([
+      'type' => 'vet_center',
```
```suggestion
      'type' => $content_type,
```

swirtSJW reviewed (changes requested):
Super minor stuff that needs resolving.

ethanteague commented:
@swirtSJW - derivatives are creating caching issues. If I move Nashville under Escanaba for the former to inherit the dashboard product treatment of the latter, the blocks created in the derivative get stuck on the term page first viewed after cache clear, so if you load Escanaba after viewing Nashville, you will see links to Nashville vc on Escanaba. 

From what I'm reading, there are a couple/few issues 
(e.g.: https://www.drupal.org/project/drupal/issues/2880682, https://www.drupal.org/project/drupal/issues/3001284)
with derivatives and how caching is handled by said api. To confirm that this behavior is associated with derivative controlled blocks, I created a new, singular block (not under purview of our derivative class), and added logic to output the vc nid associated with the current page term = correct nid is always returned.

So, I'm not sure about the best path forward. I realize that derivatives are the de facto solution for creating block multiples of similar content, but I'm not sure it's capable of handling our use case. I could pretty easily break each of these derivatives into their own block classes and place them in layout builder (probably an hour of work), but wanted to get your thoughts, first.

I updated this pr with Kaise's stuff, as well as the things you and I worked on, yesterday, and a couple of other issues I found when testing.

cc: @kevwalsh

swirtSJW commented:
@ethanteague  Interesting.  I just noticed we lost the dual view  and edit links.
spec
![image](https://user-images.githubusercontent.com/5752113/116338183-5b6b4380-a7a9-11eb-8d21-93082be48217.png)


current
![image](https://user-images.githubusercontent.com/5752113/116338273-7d64c600-a7a9-11eb-8232-219e6fc58316.png)


I don't think that was anything I touched, but I'll have to look into it.

swirtSJW commented:
I found the issue with the missing links.  They are restored now.

kevwalsh commented:
I did some clicking around from https://pr4994-ktrcnvwibjve1xd1v2cabeky6va0snm5.ci.cms.va.gov/section/veterans-health-administration/vet-centers-mvp and it looks good. 

It would be great to merge #5089 after this because it's dashboard related. 

We also spun out #5110 into its own issue. Not critical, but it's the next up that's related.

kevwalsh reviewed (approved):
QA pass!

acrollet commented:
minor note: I clicked to https://pr4994-ktrcnvwibjve1xd1v2cabeky6va0snm5.ci.cms.va.gov/section/veterans-health-administration/vet-centers-mvp and it's accessible to anonymous users - not sure if that's desired or not?

swirtSJW commented:
![image](https://user-images.githubusercontent.com/5752113/116411891-250ee200-a804-11eb-8eea-c4d17dd948bf.png)

swirtSJW commented:
@acrollet Good spotting that issue.  It is not related to this PR as the sections taxonomy is available to anon users now.   I believe that is intentional, but I don't have the history on the rationale.    Here is an example from staging.
![image](https://user-images.githubusercontent.com/5752113/116412422-ac5c5580-a804-11eb-8f92-20ca3568924b.png)

acrollet commented:
@swirtSJW good deal!

swirtSJW commented:
Will merge as soon as tests finish.
